lib/improver/spotdata/framework.py

```diff
@@ -0,0 +1,335 @@
+                  output_path=None, use_multiprocessing=False):
+    """
+    A framework that calls the components of the spotdata code. This includes
+    building site data into a suitable format, finding grid neighbours to
+    those sites with the chosen method, and then extracting data with the
+    chosen method. The final results are written out to new irregularly
+    gridded iris.cube.Cubes.
+
+    Args:
+    -----
+    config_file_path : string
+        Path to a json file defining the recipes for extracting diagnostics at
+        SpotData sites from gridded data.
+
+    data_path : string
+        Path to diagnostic data files.
+
+    ancillary_path : string
+        Path to ancillary data files.
+
+    diagnostic_list : list of strings
+        List of diagnostic names that match those defined in the config_file
+        to select which diagnostics to process.
+
+    site_path : string
+        Path to site data file if in use. If no lats/lons are specified at the
+        command line, this file path is needed.
+
+    constants_file_path : string
+        Path to a json file defining constants to be used in methods that have
+        tolerances that may be set. e.g. maxiumum vertical extrapolation/
+        interpolation of temperatures using a temperature lapse rate method.
+
+    site_properties : dict
+        Contains:
+
+        latitudes : list of ints/floats
+            A list of latitudes for running on the fly for a custom set of
+            sites. The order should correspond to the subsequent latitudes and
+            altitudes variables to construct each site.
+
+        longitudes : list of ints/floats
+            A list of longitudes for running on the fly for a custom set of
+            sites.
+
+        altitudes : list of ints/floats
+            A list of altitudes for running on the fly for a custom set of
+            sites.
+
+        site_ids : list of ints
+            A list of site_ids to associate with the above on the fly
+            constructed sites. This must be ordered the same as the latitudes/
+            longitudes/altitudes lists.
+
+    forecast_date : string (YYYYMMDD)
+        A string of format YYYYMMDD defining the start date for which forecasts
+        are required.
+
+    forecast_time : integer
+        An integer giving the hour on the forecast_date at which to start the
+        forecast output; 24hr clock such that 17 = 17Z for example.
+
+    forecast_length : integer
+        An integer giving the desired length of the forecast output in hours
+        (e.g. 48 for a two day forecast period).
+
+    output_path : string
+        Path to which output file containing processed diagnostic should be
+        written.
+
+    use_multiprocessing : boolean
+        A switch determining whether to use multiprocessing in the data
+        extraction step.
+
+    Returns:
+    --------
+    Writes out cubes of the requested diagnostics, with data extracted to the
```

Comment:
For want of a better name at present and pending a different method of using this, I'll leave it as returns.